content/en/blog/2022/frontend-overhaul/index.md

```diff
@@ -0,0 +1,182 @@
+
+We worked on an initial proposal that included the following bullet points:
+
+- Framework and tooling (Scaffolding, I/O, styling, UI library)
+- Code Architecture and structure (Directories, coding patterns)
+- Instrumentation
+- Deployment & Distribution
+- Testing (E2E, unit test)
+
+This proposal was presented to the OpenTelemetry Demo SIG duringone of the
+weekly Monday meetings and we were given the green light to move ahead. As part
+of the changes, we decided to use Next.js to not only work as the primary
+frontend application but also to work as an aggregation layer between the
+frontend and the GRPC backend services.
+
+![New Frontend Data Flow](data-flow.png)
+
+As you can see in the Diagram, the application has two major connectivity
+points, one coming from the browser side (REST) to connect to the Next.js
+aggregation layer and the other from the aggregation layer to the backend
+services (GRPC).
+
+## OpenTelemetry Instrumentation
+
+The next big thing we worked on was having a way to instrument both sides of the
+Next.js application. To accommodate this we had to connect the same application
+twice to the same open telemetry collector that was being used by all of the
+other microservices.
+
+For the backend side, a simple solution was designed that involved using the
+[official GRPC exporter](https://www.npmjs.com/package/@opentelemetry/exporter-trace-otlp-grpc)
+in combination with the
+[Nodejs SDK](https://www.npmjs.com/package/@opentelemetry/sdk-node).
+
+You can find the full
+[implementation here](https://github.com/open-telemetry/opentelemetry-demo/blob/main/src/frontend/utils/telemetry/Instrumentation.js).
+The basic instrumentation includes auto instrumentation for most of the commonly
+used
+[libraries and tools for Node.js](https://www.npmjs.com/package/@opentelemetry/auto-instrumentations-node).
+But as part of having a better example for users, a manual instrumentation piece
+was added in the form of a route middleware that would catch the incoming HTTP
+request and create a span based on it, including the context propagation. The
+[implementation can be found here](https://github.com/open-telemetry/opentelemetry-demo/blob/main/src/frontend/utils/telemetry/InstrumentationMiddleware.ts).
+
+For the frontend side, it became a little bit tricky, as the
+[initial Next.js render](https://nextjs.org/learn/foundations/how-nextjs-works/rendering)
+comes from the server side so, in this case, we had to make sure to load the
+tracer from the browser side when the Javascript code is executed.
+
+After adding validations to check for the browser side, we then can proceed to
+load the custom front-end tracing module which includes the creation of the
+[web tracer provider and the automatic web instrumentations](https://github.com/open-telemetry/opentelemetry-demo/blob/main/src/frontend/utils/telemetry/FrontendTracer.ts).
+The automatic web instrumentations capture the most common user interactions
```

Comment:
Rewording:
“The automatic front end instrumentation captures the most common user actions such as clicks, fetch requests, and page loads. In order to allow the browser side to interact with the collector, a config change is needed: enable incoming CORS requests from the web app. The resulting collector receivers looks something like this:”